### PR TITLE
Feat/create dream

### PR DESCRIPTION
**Pull request by AKhoo**
Completed basic implementation of Create Dream UI

### CODE REVIEW

File: src/FormInput.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import Form from 'react-bootstrap/Form';
+
+const FormInput = (props) => {
+  const {controlId, label, type, placeholder, subText, isRequired, value, handleChange} = props;
+  return (
+    <React.Fragment>
```
Comment (Dawenster) [inline src/FormInput.js]: Why is this necessary?

File: src/FormTextarea.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+const FormTextarea = (props) => {
+  const {controlId, label, rows, placeholder, subText, isRequired, value, handleChange} = props;
+  return (
+    <React.Fragment>
+      <Form.Group controlId={controlId}>
+        <Form.Label>{label}</Form.Label>
+        <Form.Control 
+          as="textarea"
+          rows={rows}
```
Comment (Dawenster) [inline src/FormTextarea.js]: Might be good to have a default value here in case we forget to pass it in?

FYI, this would be a good use case for Flow.js to ensure we pass in the props we need.

File: src/FormTextarea.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import Form from 'react-bootstrap/Form';
+
+const FormTextarea = (props) => {
+  const {controlId, label, rows, placeholder, subText, isRequired, value, handleChange} = props;
+  return (
+    <React.Fragment>
+      <Form.Group controlId={controlId}>
+        <Form.Label>{label}</Form.Label>
```
Comment (Dawenster) [inline src/FormTextarea.js]: If no `label`, perhaps don't show the component at all?

E.g. `{label && <Form.Label>{label}</Form.Label>}`

File: src/FormTextarea.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        <Form.Control 
+          as="textarea"
+          rows={rows}
+          required={isRequired}
+          placeholder={placeholder} 
+          value={value}
+          onChange={(e) => handleChange(e.target.value)}
+          />
+        <Form.Text className="text-muted">
```
Comment (Dawenster) [inline src/FormTextarea.js]: Same as above - show only if `subText` exists

File: src/FormTextarea.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import Form from 'react-bootstrap/Form';
+
+const FormTextarea = (props) => {
```
Comment (Dawenster) [inline src/FormTextarea.js]: You could save having to name the function by just doing `export default (props) => {...` and dropping `export default FormTextarea;` below.

More of a style thing - up to you.

File: src/Messages.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import { connect } from 'react-redux'
+import Message from './Message';
+
+const Messages = (props) => {
+  const { messages } = props;
+  return (
+    messages.map((message, index) => (
+      <Message 
+        key={'message'+index}
```
Comment (Dawenster) [inline src/Messages.js]: `index` should suffice.

File: src/Messages.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import { connect } from 'react-redux'
+import Message from './Message';
+
+const Messages = (props) => {
+  const { messages } = props;
+  return (
+    messages.map((message, index) => (
```
Comment (Dawenster) [inline src/Messages.js]: I feel like React will show a warning if you have a return that returns more than one top-level component. I didn't get to reproduce this, but I feel like this `map` needs to be wrapped in a `Fragment` or a `<>`

File: src/actions.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
+import axios from 'axios';
+import { store } from './App';
+
+export const setElementsData = (elementsData) => {
+  return {
+    type: 'SET_ELEMENTSDATA',
```
Comment (Dawenster) [inline src/actions.js]: Without using Flow.js, one way to avoid potential typos is to set these action types as constants and import them where you need them.

E.g. `SET_ELEMENTSDATA = 'SET_ELEMENTSDATA'` and then using `SET_ELEMENTSDATA` everywhere

File: src/actions.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    payload: elementsData,
+  }
+}
+
+const setLoadState = (isLoading) => {
```
Comment (Dawenster) [inline src/actions.js]: My sense is that this should not be a separate action. Why not just set the loading state as you are setting the data? Perhaps this will make more sense with the reducer refactor I'm suggesting in `reducer.js`

File: src/reducer.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
+import { combineReducers } from 'redux'
+
+const elementsData = (state = {}, action) => {
```
Comment (Dawenster) [inline src/reducer.js]: I typically see reducers setup like this: https://redux.js.org/basics/example#reducers

I think it's better that way because:
1. For a given action, there should only be one return - a `switch` perfectly models this
2. A `switch` can also catch a default state and return `state`
3. It seems to have less repetition than what you've implemented below (specifically the `else return state` stuff)

Was there a reason for your current implementation?

File: src/CreateDream.js [hunk 11]
```diff
@@ -1,9 +1,107 @@
-import React from 'react';
+import React, { useState, useEffect } from 'react';
+import { connect } from 'react-redux'
+import Button from 'react-bootstrap/Button';
+import Image from 'react-bootstrap/Image';
+import Form from 'react-bootstrap/Form';
```
[the remark this answers is not in the document]
Comment (AKhoo) [reply src/CreateDream.js]: Oh cool! I didn't know that was the same!

File: src/CreateDream.js [hunk 12]
```diff
@@ -1,9 +1,107 @@
...
+const CreateDream = (props) => {
+  const [email, setEmail] = useState('');
+  const [selectedElement, setSelectedElement] = useState({
+    name: '',
+    id: '',
+  });
+  const [description, setDescription] = useState('');
+  const {elementsData, storeElementsData, addSuccessMessage} = props;
+
+  useEffect(() => {
```
[the remark this answers is not in the document]
Comment (AKhoo) [reply src/CreateDream.js]: Both useState and useEffect are new to React Hooks!

File: src/CreateDream.js [hunk 13]
```diff
@@ -1,9 +1,107 @@
...
+
+        <Form
+          onSubmit = {(e) => {
+            e.preventDefault();
+            postDream({
+              email,
+              element_ids: selectedElement.id,
+              description,
+            })
+              .then(() => addSuccessMessage('Dream successfully submitted. Thank you!'))
```
[the remark this answers is not in the document]
Comment (AKhoo) [reply src/CreateDream.js]: Yes and yes. Added as new asana tasks in the backlog.

File: src/FormInput.js [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+      <Form.Group controlId={controlId}>
```
[the remark this answers is not in the document]
Comment (AKhoo) [reply src/FormInput.js]: I think it's a pattern that's specific to React Boostrap: https://react-bootstrap.github.io/components/forms/

It sets the id for the input that's related to the form group to the controlId, and the form label's 'for' property to the same id: https://stackoverflow.com/questions/46776626/what-is-a-controlid-in-react-bootstrap

File: src/actions.js (hunk 8 above)
Comment (AKhoo) [reply src/actions.js]: Interesting... This is only useful if we might use a specific action type more than once, right? What is a situation where we would do that?

File: src/reducer.js (hunk 10 above)
Comment (AKhoo) [reply src/reducer.js]: I was warned once that large Switch statements can be hard to debug (i.e. if you miss a return statement or break statement), so I have been avoiding Switch statements in general.... but I do see how that would avoid repetition, and Redux's official docs are using them!

File: src/actions.js (hunk 8 above)
Comment (AKhoo) [reply src/actions.js]: Discussed:
1) Create a separate file called types.js and set constants there
2) Import into actions.js and reducer.js and use it there
